### DISCUSSION TITLE
"Traditional" scratchpads like Qtile

### DISCUSSION THREAD

**Discussion by LamprosPitsillos** [hyprwm/Hyprland#3097] [Show and tell]
```sh
#!/usr/bin/env dash

# Inferno LP 2023-08-28

windows_in(){
hyprctl clients -j | jq ".[] | select(.workspace.name == \"special:$1\" )"
}

toggle_scratchpad(){
    workspace_name="$1"
    cmd="$2"

    windows=$( windows_in "$workspace_name" )
    # If not on latest , check the edit history of this post
    if [ -z "$windows" ];then
        hyprctl dispatch "exec [workspace special:$workspace_name] $cmd"
        else
        hyprctl dispatch togglespecialworkspace "$workspace_name"
    fi
}

if [ "$1" = "terminal" ];then
    toggle_scratchpad "terminal" "kitty"
elif [ "$1" = "music_player" ];then
    toggle_scratchpad "music_player" "mpv ~/music"
elif [ "$1" = "file_manager" ];then
    toggle_scratchpad "file_manager" "vifm"
fi
```
Use in config:
```
$hyprscripts= ~/path/to/scripts
bind = $mainMod, period , exec, $hyprscripts/scratchpads terminal

```

I made a script to make more traditional scratchpads (like Qtile).
* Press X key , open scratchpad AND if app is not running in scratchpad run it.
* Pres X again close scratchpad ( app still running )

# TODO
* ~~Remove the `sleep` from the first launch, it seems that `hyprland` doesn't register the window in the special workspace before a few ms pass so the workspace doesn't open since it thinks its empty~~ solved by https://github.com/hyprwm/Hyprland/commit/870471dd96724474869d2e648b623c17df12fb57

* Add _something_ to the config so that newly launched applications in scratchpad open in the workspace behind it and not the scratch pad it self ( Help is welcomed , am still new to Hyprland! ) worked on by #3192 

Relevant #2596

**Comment by dantefromhell** [comment]
I'm curious why you decided to a have a block of `if` at the end of your script instead of another script parameter?

**Reply by LamprosPitsillos** [reply]
Now that i think about it, you are right, makes since. Only problem i can think of is possible spelling mistakes while calling the script, which shouldn't cause any issues

**Reply by dantefromhell** [reply]
Yeah possibly quoting becomes important so bash/dash properly maps script input parameters.

The [hyprpaper README](https://github.com/hyprwm/hyprpaper) has some nice examples though.

**Comment by MightyPlaza** [comment]
no need for sleep anymore

**Reply by LamprosPitsillos** [reply]
Could you point to the commit tha fixed it please?

**Reply by MightyPlaza** [reply]
870471dd96724474869d2e648b623c17df12fb57

**Reply by MightyPlaza** [reply]
no need to toggle anymore as well

**Reply by LamprosPitsillos** [reply]
Wdym?

**Reply by MightyPlaza** [reply]
https://github.com/hyprwm/Hyprland/pull/3116

**Reply by LamprosPitsillos** [reply]
Sorry but, how does this remove the need for a toggle ?
Thanks

**Reply by MightyPlaza** [reply]
now when running
`hyprctl dispatch "exec [workspace special] kitty` the special workspace opens
`hyprctl dispatch "exec [workspace special silent] kitty` the special workspace doesn't open

**Reply by LamprosPitsillos** [reply]
Yes but in any case running this will cause `kitty` to start each time , regardless of if the workspace opens. The reason for the toggle is to open `kitty` only once.

I imagine after that commit this should be changed ?
```sh
if [ -z "$windows" ];then
    hyprctl dispatch "exec [workspace special:$workspace_name] $cmd"
else
    hyprctl dispatch togglespecialworkspace "$workspace_name"
fi
```

**Comment by MightyPlaza** [comment]
why not just start everything at startup and just toggle the workspaces without any script?

**Reply by LamprosPitsillos** [reply]
It's not a bad Idea but this is a bit more flexible.
Also no resources are used if you are not using the workspace.

**Reply by dantefromhell** [reply]
> no resources are used if you are not using the workspace 

I do prefer this approach in my laptop, saves battery

**Comment by dantefromhell** [comment]
> Add something to the config so that newly launched applications in scratchpad open in the workspace behind it

According to https://wiki.hyprland.org/Configuring/Dispatchers/#special-workspace newly created floating windows will be send to the "current active real" workspace.

Is there a way to set new windows on a special workspace as floating per default? That might help implement the behaviour 🤔

**Reply by LamprosPitsillos** [reply]
https://wiki.hyprland.org/Configuring/Dispatchers/#workspace-options

Maybe toggling this rule somehow after creation of the first window?
It's not very clear to me how to enable this rule though...
Edit:

```
hyprctl dispatch workspaceopt allfloat
Or
bind=X,X,workspaceopt,allfloat
```
Edit2:
doesn't work as expected, first it needs a sleep in there to allow for the first window to open and the to toggle the rule. Also windows become floating BUT dont get droped to the workspace behind.

hyprctl version : 91e28bbe9df85e2e94fbcc0137106362aea14ab5 commit
